compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -865,46 +853,121 @@ struct DistributeContract final : OpDistributionPattern<vector::ContractionOp> {
-    int64_t numElements = shaped.getNumElements();
-    SmallVector<int64_t> flatShape(1, numElements);
-    VectorType flatVecType = VectorType::get(flatShape, accElemTy);
-    VectorValue flat = rewriter.create<vector::ShapeCastOp>(loc, flatVecType,
-                                                            localContractValue);
-
-    // Step 2: Do subgroup reduction.
-    FailureOr<VectorValue> threadReduced = doThreadReduction(
-        rewriter, lhsLayout, flat, contractOp.getKind(), reducedDims);
-    if (failed(threadReduced)) {
-      return failure();
+    NestedLayoutAttr resLayout;
+    if (auto contractRes = dyn_cast<VectorValue>(contractOp.getResult())) {
+      resLayout = dyn_cast<NestedLayoutAttr>(signature[contractRes]);
+    } else {
+      // Create a zero-d layout because we
+      // are going to add reduction dims
+      // back to handle the partial reduction
+      resLayout = NestedLayoutAttr::get(contractOp.getContext(), {}, {}, {}, {},
+                                        {}, {}, {});
     }
 
-    // Do reduction against accumulator, which needs to be done after thread
-    // reduction.
-    VectorValue unflattened = rewriter.create<vector::ShapeCastOp>(
-        loc, shaped, threadReduced.value());
-
-    if (!accVector) {
-      disAcc = rewriter.create<vector::BroadcastOp>(loc, shaped, disAcc);
+    // Shapecast to re-insert reduction dimensions as unit dims.
+    // We append the result shape with reduction dimensions as
+    // the fastest changing dimensions.
+    int64_t opRank = contractOp.getIteratorTypes().size();
+    SmallVector<int64_t> partialReducedDistributedShape(opRank * 3, 0);
+    SmallVector<int64_t> resDistShape = resLayout.getDistributedShape();
+    SmallVector<int64_t> lhsDistShape = lhsLayout.getDistributedShape();
+    for (int64_t tileGroupIdx : llvm::seq<int64_t>(3)) {
+      int64_t tileGroupOffset = tileGroupIdx * opRank;
+      for (int64_t dim : llvm::seq<int64_t>(opRank)) {
+        if (dim < resLayout.getRank()) {
+          int64_t resTileGroupOffset = tileGroupIdx * resLayout.getRank();
+          partialReducedDistributedShape[tileGroupOffset + dim] =
+              resDistShape[resTileGroupOffset + dim];
+        } else {
+          // add the thread local reduced dims at the tail.
+          partialReducedDistributedShape[tileGroupOffset + dim] = 1;
+        }
+      }
     }
 
-    // Step 3: Accumulator Reduction
-    Value accReduction = vector::makeArithReduction(
-        rewriter, loc, contractOp.getKind(), unflattened, disAcc);
-    auto accReduced = dyn_cast<VectorValue>(accReduction);
-    if (!accReduced) {
-      return failure();
-    }
+    VectorType partialReducedDistributedType =
+        VectorType::get(partialReducedDistributedShape,
+                        localContractValue.getType().getElementType());
+    Value isoRankLocalReduced = rewriter.create<vector::ShapeCastOp>(
+        loc, partialReducedDistributedType, localContractValue);
```

Comment:
can we instead do a broadcast + transpose? broadcast + transpose is prefered for unit dimensions because it is easier to unroll / reason about